### ISSUE TITLE
[learning] Train a learned masked selector over fixed finite candidate tables

### ISSUE THREAD

**Issue by JanDuchscherer104** [JanDuchscherer104/ARIA-NBV#75]
Parent: #67

Depends on: #74 for online collection; validated replay/Q_H source contracts from PR #62 follow-ups.

## Outcome

Implement a learned actor-visible policy that scores and selects among the existing finite candidate rows while preserving the hard action mask, exact candidate identity, target conditioning, history/budget state, and matched hard-oracle endpoint evaluation.

This is the lowest-risk learned-control step. It uses the renderer and target-RRI scorer as a black-box teacher/environment; neither candidate generation nor the oracle needs to be differentiable.

## Current repository state

- `CounterfactualPoseGenerator` already accepts a `score_candidates` callback.
- The replay engine stores selection scores, logits, probabilities, log probabilities, entropy, selected index, and seed.
- Built-in policies include random-valid, oracle-greedy, and temperature-softmax.
- The rollout store and derived `q_h/` view contain masks, selected transitions, rewards, successor linkage, budget/horizon, and terminal state.
- The current one-step VIN path is not yet a positive-width observed-target finite-horizon policy.
- The planned target-conditioned `Q_H` model/learner remains incomplete.
- Existing rollout records detach tensors for persistence, which is correct for storage but cannot by itself support same-step online backpropagation.

Relevant owners:

- `aria_nbv/aria_nbv/rollouts/replay/engine.py`
- `aria_nbv/aria_nbv/rollouts/replay/policy.py`
- `aria_nbv/aria_nbv/data_handling/qh_data/`
- `aria_nbv/aria_nbv/vin/models/target_myopic.py`
- `aria_nbv/aria_nbv/vin/models/target_finite_horizon.py`
- Lightning/model-input contracts
- rollout/Q_H store readers

## Theory

For actor state `s_t`, target `e`, candidate table `Q_t`, and hard mask `m_t`, the policy must assign no probability to invalid rows:

```math
\pi_\theta(i\mid s_t,e,Q_t,m_t)
= \frac{m_{t,i}\exp(z_{t,i})}
       {\sum_j m_{t,j}\exp(z_{t,j})}.
```

A deterministic selector uses masked `argmax`. A stochastic behavior policy samples from the masked distribution and records its exact log probability.

VIN-NBV's central quality-driven idea is to “predict the reconstruction quality improvement of views directly.” ARIA-NBV keeps that dense one-step teacher signal but adds target conditioning, selected history, remaining budget, and finite-horizon value.

The selector and proposal generator are different learning problems:

- **selector:** ranks rows supplied by a fixed candidate generator;
- **proposer:** changes which rows exist and is handled separately.

## Required policy interface

```python
class FiniteCandidatePolicy(Protocol):
    def decide(
        self,
        observation: ActorCandidateObservation,
        *,
        stochastic: bool,
        generator: torch.Generator | None,
    ) -> PolicyDecision: ...

@dataclass
class PolicyDecision:
    values: Tensor                 # [N]
    logits: Tensor                 # [N]
    probabilities: Tensor          # [N]
    log_probabilities: Tensor      # [N]
    selected_shell_index: int
    selected_valid_index: int
    selected_log_probability: Tensor
    entropy: Tensor
    checkpoint_hash: str
    model_config_hash: str
    behavior_policy_id: str
```

Keep a live gradient-bearing decision object inside the collector/learner. Create a separate detached persistence DTO at the storage boundary.

## Actor input contract

The model may consume:

```text
actor-visible scene carrier
observed/predicted target descriptor and source mask
current/reference-relative candidate geometry
candidate-target relation/framing features
selected-pose/directional history
step, requested horizon, and remaining budget
candidate family/proposal provenance
hard validity/reason fields as masked support information
```

It must not consume:

```text
GT mesh or target crop
all-candidate rendered depth/points
oracle RRI or target errors
future selected observations
GT identity/match fields except as labels/audit
```

### Stage 1 — actor-visible learned myopic selector

Train one value per valid candidate from dense one-step target labels:

```math
\mathcal L_{1} =
\frac{\sum_i m^{train}_i\,\ell(\hat r_i,r_i)}
     {\sum_i m^{train}_i+\epsilon}.
```

Use continuous root-normalized target gain as the primary regression/ranking signal. CORAL may remain a matched one-step control, but finite-horizon return must remain continuous and additive.

Required controls:

- masked independent MLP;
- candidate-to-state cross-attention;
- optional DeepSets/Set Transformer only after row/mask invariants pass.

### Stage 2 — offline finite-horizon `Q_H`

Implement masked Double-Q backups over selected transitions:

```math
y_t^{(h)} = r_t + \gamma(1-d_t)
Q^-_{h-1}\!\left(s_{t+1},
\arg\max_{i:m_{t+1,i}=1}Q_{h-1}(s_{t+1},i)\right).
```

Requirements:

- `Q_0 = 0`;
- no valid successor means terminal bootstrap zero;
- action selection, target selection, and loss use the correct masks;
- no invented labels for unselected transitions;
- requested horizon and remaining budget are explicit inputs.

### Stage 3 — optional online fine-tuning

After offline calibration and endpoint evaluation pass, collect transitions through #74:

1. policy sees actor observation and fixed candidate table;
2. hard mask is applied;
3. action is selected/sampled;
4. oracle evaluates selected action or dense table;
5. transition enters replay with behavior checkpoint and log probability;
6. learner updates asynchronously or between bounded collection phases.

Candidate generation remains fixed during this stage, so improvements are attributable to selection.

Possible updates:

- continued supervised learning from dense oracle labels;
- advantage-weighted behavior cloning from replay;
- policy-gradient/actor-critic updates from selected-only rewards;
- conservative offline updates when mixing old behavior policies.

Start with dense supervised/Double-Q training before PPO-style on-policy optimization; the oracle already supplies more information than a single sampled reward.

## Selection modes

Add explicit policy enum values rather than overloading oracle modes:

```text
LEARNED_GREEDY
LEARNED_SOFTMAX
LEARNED_EPSILON_GREEDY
LEARNED_QH_GREEDY
LEARNED_QH_SOFTMAX
```

Every selection record must state whether `values` are immediate gain, `Q_H`, logits, or a calibrated ranking proxy.

## Provenance

Persist per transition:

```text
policy architecture/version
checkpoint content hash
model config hash
training data/manifest hash
behavior policy mode
temperature/epsilon
candidate table hash
selected log probability
entropy
value/logit vectors or retained compact form
online update/collection epoch
```

A checkpoint filename is not sufficient provenance.

## Evaluation protocol

On frozen scene-disjoint roots and candidate tables, compare:

```text
random valid
actor-visible learned myopic
oracle one-step greedy
bounded oracle lookahead
learned Q_H
```

All policies receive equal acquisition budget and matched candidate support. Final policy quality is independent endpoint target gain from the hard oracle, not training loss or predicted value.

Primary policy metrics:

- endpoint target gain;
- scene-paired difference;
- oracle-lookahead headroom and recovered fraction;
- selected-action oracle regret;
- early termination/no-action rate;
- path cost and invalidity;
- family and directional selection diversity;
- runtime and calibration.

## Acceptance criteria

- [ ] A typed finite-candidate policy interface returns live and persisted decision records.
- [ ] Learned logits/probabilities are exactly zero/`-inf` outside the hard mask.
- [ ] A positive-width observed/predicted target path is implemented and leakage-tested.
- [ ] The learned myopic selector trains from dense valid-row target labels.
- [ ] A masked continuous-return `Q_H` learner trains from selected-transition replay.
- [ ] Row permutation equivariance, invalid-row isolation, duplicate-row, and valid-count tests pass.
- [ ] Every behavior decision is bound to checkpoint/config/candidate-table provenance.
- [ ] The online collector can use dense, subset, or selected-only oracle queries from #74.
- [ ] Candidate generation remains frozen for selector experiments.
- [ ] Held-out selected trajectories are independently re-evaluated by the hard endpoint oracle.

## Required tests

- masked softmax and masked argmax edge cases;
- all-invalid/no-successor termination;
- row permutation equivariance;
- invalid and padded row perturbation isolation;
- duplicate-row and valid-count invariance;
- target-source dropout/leakage;
- horizon/budget boundary and `Q_0`;
- Double-Q online-selector/target-evaluator separation;
- live-gradient versus detached persistence boundary;
- checkpoint/content-hash mismatch rejection;
- deterministic stochastic decisions under a keyed generator;
- matched candidate-table equality across policies;
- endpoint oracle re-evaluation.

## External sources

- Frahm et al., *VIN-NBV*: https://arxiv.org/abs/2505.06219
- van Hasselt et al., *Deep Reinforcement Learning with Double Q-learning*: https://arxiv.org/abs/1509.06461
- Kumar et al., *Conservative Q-Learning*: https://proceedings.neurips.cc/paper/2020/hash/0d2b2061826a5df3221116a5085a6052-Abstract.html

## Non-goals

- This issue does not learn candidate poses or family counts.
- It does not require differentiating through rendering, validity, fusion, or RRI.
- It does not expose oracle candidate renders to the actor.
- It does not claim policy quality from train-only collection or loss curves.

**Comment by JanDuchscherer104** [comment]
## Research enrichment — round 2: support shift, provenance shortcuts, and mask realism

### 1. Treat candidate-profile shift as a first-class generalization axis

A selector trained on one fixed family allocation can learn the proposal distribution rather than geometric utility. Family ID, component block, row location, seed/replica, and profile ID may correlate with oracle gain.

Geirhos et al. describe shortcut learning as reliance on rules that perform on standard benchmarks but fail under changed conditions:

- https://doi.org/10.1038/s42256-020-00257-z

Add the following controls:

```text
provenance-feature dropout
family/profile ID removal
component-block and row shuffling
same physical candidate with benignly changed provenance encoding
train on several profiles, test on a held-out profile
train on fixed-attempt shells, test on reservoir-selected shells and vice versa
unseen family with overlapping physical pose distribution
```

Report within-profile and cross-profile selected-action regret separately.

### 2. Distinguish three masks

The current hard mask is generated from privileged GT geometry. That is valid for offline oracle experiments, but a deployable actor cannot generally know it. Keep:

```text
oracle_action_mask
actor_observed_action_mask
learned_feasibility_probability / abstention
```

Evaluate policy modes:

1. `ORACLE_MASK_UPPER_BOUND` — actor values with privileged hard mask;
2. `ACTOR_MASK` — only observed free/occupied/unknown evidence;
3. `LEARNED_FEASIBILITY` — calibrated prediction with hard known constraints;
4. `UNMASKED_DIAGNOSTIC` — never deployed; measures reliance on invalid rows.

The thesis hard-mask result must state which mask source it uses. Do not infer deployment safety from oracle-mask performance.

### 3. Add conditional value and feasibility factorization

Retain the planned structure:

```math
p_\theta(valid\mid s,e,q),
\qquad
Q_\theta(s,e,q\mid valid).
```

Invalid rows supervise feasibility only; RRI/value loss remains valid-row-only. Evaluate:

```text
Brier score / ECE for feasibility
false-safe rate
risk-coverage or abstention curve
conditional value calibration
validity-times-value ranking as a named ablation
```

The hard available mask remains authoritative where known.

### 4. Add decision-focused metrics beyond global rank correlation

Spearman correlation can improve while the selected top action remains poor. Report per state and scene:

```text
top-1 oracle regret
top-k oracle recall
NDCG or tolerance-aware top-set recall
selected target gain
calibration of predicted gain
policy endpoint gain
```

Use the oracle sensitivity/noise tolerance from #80 to avoid penalizing arbitrary ordering among indistinguishable near-ties.

### 5. Audit offline support and extrapolation

BCQ and CQL are relevant because value maximization can exploit unsupported actions in offline replay:

- BCQ: https://proceedings.mlr.press/v97/fujimoto19a.html
- CQL: https://proceedings.neurips.cc/paper/2020/hash/0d2b2061826a5df3221116a5085a6052-Abstract.html

For ARIA-NBV, support includes both state and candidate-action context. Persist/query:

```text
behavior policy/proposal identity
selected-action probability
candidate profile and replica
state acquisition mode/round (#81)
nearest replay support / OOD score
```

Add conservative/OOD controls only after the supervised myopic and factual Double-Q baselines work. A high predicted `Q_H` on a rare/unseen candidate profile must trigger support diagnostics and hard-oracle re-evaluation.

### 6. Use policy-induced data only through explicit aggregation rounds

DAgger motivates adding learner-visited states to training:

https://proceedings.mlr.press/v15/ross11a.html

Use #74 and #81 to collect train-scene learner-induced states with exact checkpoint/proposal/query provenance. Keep the held-out evaluation population untouched. Compare:

```text
frozen offline only
one or more bounded DAgger-style aggregation rounds
```

### 7. Add counterfactual state-fidelity ablations

Train/evaluate separately under #74's successor modes:

```text
pose/history only
privileged selected GT depth
actor-belief/noisy selected observation
```

A model that succeeds only with perfect selected GT geometry has established an oracle-state result, not a deployable actor-state result.

### 8. Prevent set-context leakage from invalid/padded rows

Beyond perturbation tests, add gradients and normalization checks:

```text
invalid-row feature gradients are zero into valid-row outputs
adding duplicate invalid rows does not change valid scores
changing valid count only affects declared valid-count/support inputs
set normalization excludes masked/padded rows
```

### Additional acceptance criteria

- [ ] Candidate-profile holdout and provenance-dropout results are required for selector promotion.
- [ ] Oracle, actor, and learned feasibility masks are distinct inputs/evaluation modes.
- [ ] Feasibility calibration and false-safe risk are reported.
- [ ] Selected-action regret and endpoint gain complement Spearman/ordinal metrics.
- [ ] Offline support/OOD evidence is bound to state, candidate profile, and behavior provenance.
- [ ] DAgger-style data aggregation is train-only and round-versioned.
- [ ] Policy claims identify successor-state fidelity.
- [ ] Invalid/padded rows cannot affect valid-row values through attention or normalization unless an explicit support feature does so.

**Comment by JanDuchscherer104** [comment]
## Scorer-stack audit update — dense labels are present, but the current learner remains selected-action-only

Live-stack evidence from PRs #129–#138 narrows the remaining objective gap:

- `qh_dense_valid_fitted_q_v1` correctly requires `label_mask == action_mask` on every realized state, so all hard-valid rows carry finite one-step target-root-gain labels.
- `QhLightningModule._fitted_q_admission_mask`, however, is still based on `selected_train_mask`.
- `_fitted_q_components` gathers `conditional_q` only at the factual selected index and computes one value loss per state.
- The successor maximum may then select any densely labelled valid row, even though unselected rows received no direct Q1 loss.

Thus the new corpus is dense-valid, but the fitted learner does **not yet use the available dense Q1 supervision**. This wastes the strongest label surface and makes learned exact-Q2 error partly an avoidable Q1 support/extrapolation problem.

A second identifiability constraint remains: all training calls currently use `requested_horizon = horizon_remaining`. Therefore the model is trained only on diagonal pairs `(b_t,h)=(b_t,b_t)` even though its public scorer accepts `1 <= h <= b_t`. Queries with `h < b_t` are presently unsupported extrapolation.

### Smallest attributable objective repair

1. For `h=1`, train `conditional_q[..., i]` on every row admitted by `q_train_mask` using the dense immediate target.
2. For `h>1`, retain factual selected-action TD supervision unless another action has a materialized successor.
3. Keep hard validity, candidate materialization, and label support as separate masks; do not turn invalid rows into low Q targets.
4. Either restrict V1 inference to `h == b_t`, or materialize/sample genuine off-diagonal horizon targets with exact `h -> h-1` recursion.
5. Freeze the dense-Q1 versus selected-TD weighting/sampling rule so Q1 volume cannot hide longer-horizon failure.

### Evidence gate

Report per horizon and separately for factual-selected versus unselected valid rows:

```text
labelled row count and scene/target support
MAE / Huber loss and rank correlation
top-1 oracle regret
successor-max calibration
exact-Q2 MAE and support
(horizon_remaining, requested_horizon) support matrix
```

Promotion should require that every inference-admitted `(b_t,h)` pair appears in the training-support receipt. This is independent of the existing endpoint-policy gate and does not weaken the hard action mask.
